pkg/compact/compact.go

```diff
@@ -470,6 +502,60 @@ func (cg *Group) Resolution() int64 {
 	return cg.resolution
 }
 
+// should return the results/metrics of the planning simulation
+type PlanSim interface {
+	ProgressCalculate(ctx context.Context, groups []*Group) error
+}
+
+type DefaultPlanSim struct {
+	planner Planner
+}
+
+func (ps *DefaultPlanSim) ProgressCalculate(ctx context.Context, groups []*Group) error {
+	iterations := 0
+	blocksToMerge := 0
+
+	// figure out hasPlan and noPlan
+	for {
+		for _, g := range groups {
+			if len(g.IDs()) == 1 {
+				continue
+			}
+
+			// parameter should be of type tsdb.BlockMeta.meta
+			plan, err := ps.planner.Plan(ctx, g.metasByMinTime)
+			if err != nil {
+				return errors.Wrapf(err, "could not plan")
+			}
+			if len(plan) == 0 {
+				continue
+			}
+			iterations++
+
+			var toRemove map[ulid.ULID]bool
```

Comment:
So this is a set and the size is still `len(plan)`. You can just use `make(map[ulid.ULID]struct{}, len(plan))` to initialize it.

Reply:
I'll initialise it with len(plan) but I don't get why using a struct is better than using a bool here?

Reply:
struct{} is an empty struct and takes 0 byte. https://www.davidkaya.com/sets-in-golang/